primitives/xcm-bridge-hub/src/lib.rs

```diff
@@ -0,0 +1,585 @@
+}
+
+/// Locations of bridge endpoints at both sides of the bridge.
+#[derive(Clone, RuntimeDebug, PartialEq, Eq)]
+pub struct BridgeLocations {
+	/// Relative (to this bridge hub) location of this side of the bridge.
+	pub bridge_origin_relative_location: MultiLocation,
+	/// Universal (unique) location of this side of the bridge.
+	pub bridge_origin_universal_location: InteriorMultiLocation,
+	/// Relative (to this bridge hub) location of the other side of the bridge.
+	pub bridge_destination_relative_location: MultiLocation,
+	/// Universal (unique) location of other side of the bridge.
+	pub bridge_destination_universal_location: InteriorMultiLocation,
+	/// An identifier of the dedicated bridge message lane.
+	pub lane_id: LaneId,
+}
+
+/// Errors that may happen when we check bridge locations.
+#[derive(Encode, Decode, RuntimeDebug, PartialEq, Eq, PalletError, TypeInfo)]
+pub enum BridgeLocationsError {
+	/// Origin or destination locations are not universal.
+	NonUniversalLocation,
+	/// Bridge origin location is not supported.
+	InvalidBridgeOrigin,
+	/// Bridge destination is not supported (in general).
+	InvalidBridgeDestination,
+	/// Destination location is within the same global consensus.
+	DestinationIsLocal,
+	/// Destination network is not the network we are bridged with.
+	UnreachableDestination,
+	/// Destination location is unsupported. We only support bridges with relay
+	/// chain or its parachains.
+	UnsupportedDestinationLocation,
+}
+
+/// Given XCM locations, generate lane id and universal locations of bridge endpoints.
+///
+/// The `here_universal_location` is the universal location of the bridge hub runtime.
+///
+/// The `bridge_origin_relative_location` is the relative (to the `here_universal_location`)
+/// location of the bridge endpoint at this side of the bridge. It may be the parent relay
+/// chain or the sibling parachain. All junctions below parachain level are dropped.
+///
+/// The `bridge_destination_universal_location` is the universal location of the bridge
+/// destination. It may be the parent relay or the sibling parachain of the **bridged**
+/// bridge hub. All junctions below parachain level are dropped.
+///
+/// Why we drop all junctions between parachain level - that's because the lane is a bridge
+/// between two chains. All routing under this level happens when the message is delivered
+/// to the bridge destination. So at bridge level we don't care about low level junctions.
+///
+/// Returns error if `bridge_origin_relative_location` is outside of `here_universal_location`
+/// local consensus OR if `bridge_destination_universal_location` is not a universal location.
+pub fn bridge_locations(
+	here_universal_location: Box<InteriorMultiLocation>,
+	bridge_origin_relative_location: Box<MultiLocation>,
+	bridge_destination_relative_location: Box<MultiLocation>,
+	expected_remote_network: NetworkId,
+) -> Result<Box<BridgeLocations>, BridgeLocationsError> {
+	fn strip_low_level_junctions(
```

Comment:
I would change this to something like:

```
fn split_up_to_parachain(
	location: InteriorMultiLocation,
) -> Result<InteriorMultiLocation, BridgeLocationsError> {
	let mut junctions = location.into_iter();

	let global_consensus = junctions
		.next()
		.filter(|junction| matches!(junction, GlobalConsensus(_)))
		.ok_or(BridgeLocationsError::NonUniversalLocation)?;

	let maybe_parachain = junctions.next().filter(|junction| matches!(junction, Parachain(_)));
	Ok(match maybe_parachain {
		Some(parachain) => X2(global_consensus, parachain),
		None => X1(global_consensus),
	})
}
```

And would define it as a normal function (not an imbricated one).

Reply:
I've left it nested - don't want to spoil the outside module with it :)